tests/test_agent_manager.py

```diff
@@ -1296,3 +1298,92 @@ def log_contains_debug_line():
         return "DEBUG    inmanta.protocol.endpoints Start transport for client agent" in log_content
 
     await retry_limited(log_contains_debug_line, 10)
+
+
+async def test_heartbeat_different_session(server_pre_start, async_finalizer, caplog):
+    """
+    Verify that:
+      - if the max_clients is reached, the heartbeat will still work as it is in a different pool
+      - the max_clients option in the config changes the number of max_clients in a pool
+      - debug logs concerning 'max_clients limit reached' logged by Tornado, are logged as inmanta warnings.
+    """
+    caplog.set_level(logging.WARNING)
+    hanglock = asyncio.Event()
+
+    @typedmethod(
+        path="/test",
+        operation="GET",
+        client_types=["agent"],
+        agent_server=True,
+    )
+    def test_method(number: int) -> ReturnValue[int]:  # NOQA
+        """
+        api endpoint that never returns
+        """
+
+    class TestSlice(ServerSlice):
+        @handle(test_method)
+        async def test_method_implementation(self, number: int) -> ReturnValue[int]:  # NOQA
+            LOGGER.warning(f"HANG {number}")
+            await hanglock.wait()
+            return ReturnValue(response=number)
+
+    Config.set("agent_rest_transport", "max_clients", "1")
+
+    server = TestSlice(name="test_slice")
+
+    ibl = InmantaBootloader()
+    ctx = ibl.load_slices()
+
+    for mypart in ctx.get_slices():
+        ibl.restserver.add_slice(mypart)
+
+    ibl.restserver.add_slice(server)
+
+    await ibl.start()
+    async_finalizer.add(server.stop)
+    async_finalizer.add(ibl.stop)
+
+    client = protocol.Client("client")
+
+    result = await client.create_project("project-test")
+    assert result.code == 200
+    proj_id = result.result["project"]["id"]
+
+    result = await client.create_environment(proj_id, "test", None, None)
+    assert result.code == 200
+    env_id = result.result["environment"]["id"]
+    environment = await data.Environment.get_by_id(uuid.UUID(env_id))
+
+    agent_manager = ibl.restserver.get_slice(SLICE_AGENT_MANAGER)
+
+    a = Agent(hostname="node1", environment=environment.id, agent_map={"agent1": "localhost"}, code_loader=False)
+    await a.add_end_point_name("agent1")
+    async_finalizer.add(a.stop)
+    await a.start()
+
+    # Wait until session is created
+    await retry_limited(lambda: len(agent_manager.sessions) == 1, 10)
+
+    # Have many connections in flight
+    hangers = asyncio.gather(*(a._client.test_method(i) for i in range(5)))
+    logging.warning("WAITING!")
+    await asyncio.sleep(1)
+    assert not hangers.done()
+
+    def did_exceed_capacity():
+        return "WARNING max_clients limit reached, request queued. 1 active, 4 queued requests." in caplog.text
```

Comment:
I would maybe replace this function with `log_contains(caplog, <logger>, <logger_level>, <msg>)` to assert the right logger is being used

Reply:
unfortunately log_contains can not be used with retry_limited as it will assert false which will break out of the test :(

Reply:
Ah right, then maybe iterate over `caplog.get_records(...)` as it is done in log_contains to check that the message is logged as an inmanta warning and not as a tornado debug.